### PR TITLE
Add alert rules to argo-controller based on the KF093 spec

### PR DESCRIPTION
**Pull request by rgildein**
These alert rules provide an overview of all service states.

Using the KubeflowServiceDown or KubeflowServiceIsNotStable filter, the user
can easily see the status of all Kubeflow services.

These changes can be tested by running the following commands:
```bash
tox -e integration -- --keep-models
juju -m <model-name> show-unit grafana-agent-k8s/0 --endpoint metrics-endpoint | yq '.[]."relation-info".[]."application-data".alert_rules | fromjson'
# if you have cos deployed
juju consume <controller>:cos.remote-write
juju integrate remote-write grafana-agent-k8s
# open Grafana UI and check if KubeflowServiceDown and KubeflowServiceIsNotStable are there
# you can even stop pebble service to see alert rules firing
kubectl exec -it -n <model-name> pod/<pod-name> -c <workload-container> -- /charm/bin/pebble start <service-name>
```

part-of: [#1026](https://github.com/canonical/bundle-kubeflow/issues/1026)

### CODE REVIEW

File: charms/argo-controller/src/prometheus_alert_rules/KubeflowArgoControllerServices.rules [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  rules:
+  - alert: KubeflowServiceDown
+    expr: up{} < 1
+    for: 5m
+    labels:
+      severity: critical
+    annotations:
+      summary: "{{ $labels.juju_charm }} service is Down ({{ $labels.juju_model }}/{{ $labels.juju_unit }})"
+      description: |
+       One or more targets of {{ $labels.juju_charm }} charm are down on unit {{ $labels.juju_model }}/{{ $labels.juju_unit }}.
```
Comment (misohu) [inline charms/argo-controller/src/prometheus_alert_rules/KubeflowArgoControllerServices.rules:12]: Is this description correct ? I believe this alert rule fires if less then 1  target is up right ?

**Review by misohu — COMMENTED** [review]
I think we are missing tests for this change. I believe the in the issue https://github.com/canonical/bundle-kubeflow/issues/1026 we are talking about using this https://github.com/canonical/charmed-kubeflow-chisme/blob/b64f6ccca228d08c6533cbe83382306cd46670dc/src/charmed_kubeflow_chisme/testing/cos_integration.py#L413 for testing.

Can you also please provide minimal steps for reviewer to test it locally? What should I deploy and what should I check.

**Comment by rgildein** [comment]
> I think we are missing tests for this change. I believe the in the issue [canonical/bundle-kubeflow#1026](https://github.com/canonical/bundle-kubeflow/issues/1026) we are talking about using this https://github.com/canonical/charmed-kubeflow-chisme/blob/b64f6ccca228d08c6533cbe83382306cd46670dc/src/charmed_kubeflow_chisme/testing/cos_integration.py#L413 for testing.

That's the recommended way. I created this and other PRs like this by script, so I did not check if charm implement the testing or not. That means that for some charms (like this one) we only update the alert rule and for some we add it. Do you think that we should include those test to all of them?

> Can you also please provide minimal steps for reviewer to test it locally? What should I deploy and what should I check.

Something like deploy grafana-agent and check that alert rule is part of relation or deploy whole cos and check it in Grafana UI? WDYT?

File: charms/argo-controller/src/prometheus_alert_rules/KubeflowArgoControllerServices.rules (hunk 1 above)
Comment (rgildein) [reply charms/argo-controller/src/prometheus_alert_rules/KubeflowArgoControllerServices.rules:12]: The up represent the state of metrics scraping. I'll try to explain it with example:
For charm with two target, we can have three scenarios:
- both targets are working fine -> up{juju_charm="my-charm"} == 1
- one target is up and one down -> up{juju_charm="my-charm"} == 0
- both targets are down -> up{juju_charm="my-charm"} == 0

**Review by misohu — APPROVED** [review]
LGTM :)

**Pull request merged**